controllers/namespacescope_controller.go

```diff
@@ -439,19 +405,172 @@ func (r *NamespaceScopeReconciler) CreateUpdateRoleBinding(labels map[string]str
+		if err != nil {
+			return err
+		}
+
+		if err := r.CreateRole(roleList, labels, fromNs, toNs); err != nil {
+			if errors.IsForbidden(err) {
+				r.Recorder.Eventf(instance, corev1.EventTypeWarning, "Forbidden", "cannot create resource roles in API group rbac.authorization.k8s.io in the namespace %s", toNs)
+			}
+			return err
+		}
+		if err := r.CreateRoleBinding(roleList, labels, sa, fromNs, toNs); err != nil {
+			if errors.IsForbidden(err) {
+				r.Recorder.Eventf(instance, corev1.EventTypeWarning, "Forbidden", "cannot create resource rolebindings in API group rbac.authorization.k8s.io in the namespace %s", toNs)
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *NamespaceScopeReconciler) GetServiceAccountFromNamespace(instance *operatorv1.NamespaceScope, namespace string) ([]string, error) {
+	labels := instance.Spec.RestartLabels
+	pods := &corev1.PodList{}
+	opts := []client.ListOption{
+		client.MatchingLabels(labels),
+		client.InNamespace(namespace),
+	}
+
+	if err := r.List(ctx, pods, opts...); err != nil {
+		klog.Errorf("Cannot list pods with labels %v in namespace %s: %v", labels, namespace, err)
+		return nil, err
+	}
+
+	// By default, set ibm-namespace-scope-operator service account
+	var saNames = []string{"ibm-namespace-scope-operator"}
+
+	for _, pod := range pods.Items {
+		if len(pod.Spec.ServiceAccountName) != 0 {
+			saNames = append(saNames, pod.Spec.ServiceAccountName)
+		}
+	}
+
+	if len(instance.Spec.ServiceAccountMembers) != 0 {
+		for _, sa := range instance.Spec.ServiceAccountMembers {
+			serviceaccount := &corev1.ServiceAccount{}
+			if err := r.Get(ctx, types.NamespacedName{Namespace: namespace, Name: sa}, serviceaccount); err != nil {
+				klog.Errorf("Failed to get service account %s in namespace %s", sa, namespace)
+				continue
+			}
+			saNames = append(saNames, sa)
+		}
+	}
+
+	saNames = util.ToStringSlice(util.MakeSet(saNames))
+
+	return saNames, nil
+}
+
+func (r *NamespaceScopeReconciler) GetRolesFromServiceAccount(sa string, namespace string) ([]string, error) {
+	roleBindings := &rbacv1.RoleBindingList{}
+	opts := []client.ListOption{
+		client.InNamespace(namespace),
+	}
+
+	if err := r.List(ctx, roleBindings, opts...); err != nil {
+		klog.Errorf("Cannot list rolebindings with in namespace %s: %v", namespace, err)
+		return nil, err
+	}
+
+	var roleNameList []string
+	for _, roleBinding := range roleBindings.Items {
+		for _, subject := range roleBinding.Subjects {
+			if subject.Name == sa && subject.Kind == "ServiceAccount" {
+				roleNameList = append(roleNameList, roleBinding.RoleRef.Name)
+			}
+		}
+	}
+
+	return util.ToStringSlice(util.MakeSet(roleNameList)), nil
+}
+
+func (r *NamespaceScopeReconciler) CreateRole(roleNames []string, labels map[string]string, fromNs string, toNs string) error {
+	for _, roleName := range roleNames {
+		originalRole := &rbacv1.Role{}
+		if err := r.Get(ctx, types.NamespacedName{Name: roleName, Namespace: fromNs}, originalRole); err != nil {
```

Comment:
we need handle Notfound error, if originalRole Notfound, continue

Reply:
Thanks, I will update to handle the NotFound error.